plugins/usi-dock/fu-usi-dock-mcu-device.c

```diff
@@ -0,0 +1,777 @@
+					 TAG_TAG2_CMD_MCU,
+					 inbuf,
+					 sizeof(inbuf),
+					 outbuf,
+					 sizeof(outbuf),
+					 error))
+		return FALSE;
+
+	for (guint i = 0; components[i].name != NULL; i++) {
+		const guint8 *val = outbuf + components[i].offset;
+		g_autofree gchar *version = NULL;
+		g_autofree gchar *instance_id = NULL;
+		g_autoptr(FuDevice) child = NULL;
+
+		child = fu_usi_dock_child_new(fu_device_get_context(FU_DEVICE(self)));
+		if (g_strcmp0(components[i].name, "bcdVersion") == 0) {
+			if ((val[0] == 0x00 && val[1] == 0x00) ||
+			    (val[0] == 0xFF && val[1] == 0xFF)) {
+				g_debug("ignoring %s", components[i].name);
+				continue;
+			}
+			if (fu_device_has_private_flag(FU_DEVICE(self),
+						       FU_USI_DOCK_DEVICE_FLAG_VERFMT_HP)) {
+				version = g_strdup_printf("%x.%x.%x.%x",
+							  val[0] >> 4,
+							  val[0] & 0xFu,
+							  val[1] >> 4,
+							  val[1] & 0xFu);
+				fu_device_set_version_format(FU_DEVICE(self),
+							     FWUPD_VERSION_FORMAT_QUAD);
+				fu_device_set_version(FU_DEVICE(self), version);
+			} else {
+				version = g_strdup_printf("%x.%x.%02x",
+							  val[0] & 0xFu,
+							  val[0] >> 4,
+							  val[1]);
+				g_debug("ignoring %s --> %s", components[i].name, version);
+			}
+			continue;
+		} else if (g_strcmp0(components[i].name, "DMC") == 0) {
+			if ((val[2] == 0x00 && val[3] == 0x00 && val[4] == 0x00) ||
+			    (val[2] == 0xFF && val[3] == 0xFF && val[4] == 0xFF)) {
+				g_debug("ignoring %s", components[i].name);
+				continue;
+			}
+			version = g_strdup_printf("%d.%d.%d", val[2], val[3], val[4]);
+			fu_device_set_version_format(child, FWUPD_VERSION_FORMAT_TRIPLET);
+			fu_device_set_version(child, version);
+			fu_device_set_name(child, "Dock Management Controller");
+		} else if (g_strcmp0(components[i].name, "PD") == 0) {
+			if ((val[1] == 0x00 && val[2] == 0x00 && val[3] == 0x00 &&
+			     val[4] == 0x00) ||
+			    (val[1] == 0xFF && val[2] == 0xFF && val[3] == 0xFF &&
+			     val[4] == 0xFF)) {
+				g_debug("ignoring %s", components[i].name);
+				continue;
+			}
+			if (fu_device_has_private_flag(FU_DEVICE(self),
+						       FU_USI_DOCK_DEVICE_FLAG_VERFMT_HP)) {
+				version =
+				    g_strdup_printf("%d.%d.%d.%d", val[3], val[4], val[1], val[2]);
+				fu_device_set_version_format(child, FWUPD_VERSION_FORMAT_QUAD);
+			} else {
+				version = g_strdup_printf("%d.%d.%d", val[2], val[3], val[4]);
+				fu_device_set_version_format(child, FWUPD_VERSION_FORMAT_TRIPLET);
+			}
+			fu_device_set_version(child, version);
+			fu_device_set_name(child, "Power Delivery");
+		} else if (g_strcmp0(components[i].name, "TBT4") == 0) {
+			if ((val[1] == 0x00 && val[2] == 0x00 && val[3] == 0x00) ||
+			    (val[1] == 0xFF && val[2] == 0xFF && val[3] == 0xFF)) {
+				g_debug("ignoring %s", components[i].name);
+				continue;
+			}
+			version = g_strdup_printf("%02x.%02x.%02x", val[1], val[2], val[3]);
+			fu_device_set_version_format(child, FWUPD_VERSION_FORMAT_TRIPLET);
```

Comment:
really?  This isn't a pair?  I've never seen a thunderbolt device with a triplet.

Reply:
Internal defined for display only.